thorlcr/master/thmastermain.cpp

```diff
@@ -196,70 +196,68 @@ class CRegistryServer : public CSimpleInterface
             watchdog->addSlave(ep);
         ++slavesRegistered;
     }
-    bool connect()
+    bool connect(unsigned slaves)
     {
-        unsigned slaves = queryNodeClusterWidth();
         LOG(MCdebugProgress, thorJob, "Waiting for %d slaves to register", slaves);
-        unsigned timeWaited = 0;
-        unsigned connected = 0;
-        Owned<IBitSet> connectedSet = createThreadSafeBitSet();
-        for (;;)
+
+        IPointerArrayOf<INode> connectedSlaves;
+        connectedSlaves.ensure(slaves);
+        unsigned remaining = slaves;
+        INode *_sender = nullptr;
+        CMessageBuffer msg;
+        while (remaining)
         {
-            CTimeMon tm(msgDelay);
-            UnsignedArray todo;
-            unsigned s = 0;
-            while ((s=connectedSet->scan(s, false)) < slaves)
-                todo.append(s++);
-            Owned<IShuffledIterator> shuffled = createShuffledIterator(todo.ordinality());
-            ForEach(*shuffled)
+            if (!queryWorldCommunicator().recv(msg, nullptr, MPTAG_THORREGISTRATION, &_sender, MP_WAIT_FOREVER))
             {
-                s = todo.item(shuffled->get());
-                unsigned remaining;
-                if (tm.timedout(&remaining))
-                    break;
-                PROGLOG("Verifying connection to slave %d", s+1);
-                if (queryWorldCommunicator().verifyConnection(&queryNodeGroup().queryNode(s+1), remaining))
-                {
-                    StringBuffer str;
-                    PROGLOG("verified connection with %s", queryNodeGroup().queryNode(s+1).endpoint().getUrlStr(str.clear()).str());
-                    ++connected;
-                    connectedSet->set(s);
-                }
-                if (stopped)
-                    return false;
+                ::Release(_sender);
+                PROGLOG("Failed to initialize slaves");
+                return false;
+            }
+            Owned<INode> sender = _sender;
+            if (NotFound != connectedSlaves.find(sender))
+            {
+                StringBuffer epStr;
+                PROGLOG("Same slave registered twice!! : %s", sender->endpoint().getUrlStr(epStr).str());
+                return false;
+            }
+
+            /* NB: in base metal setup, the slaves know which slave they in advance, and send their slavenum at registration.
+             * In container setup, they do not send a slave by default and instead are given a # once all are registered
```

Comment:
picky(!):  non attached storage mode rather than container mode.  We will probably use containers with attached storage.  Similar comments elsewhere about bare metal should be checked.

Reply:
ok will change comment